### ISSUE TITLE
Position attribute

### ISSUE THREAD

**Issue by 1rch** [diarmidmackenzie/instanced-mesh#9]
Hi Diarmid, hope you're doing well.

Just dropping by to ask if there's any particular reason why changing the default position attribute in A-Frame doesn't change the position of the instanced meshes dynamically? I see you're using a separate component for moving the meshes, but it seems kind of counter-intuitive to how I'm used to working with A-Frame. 
What are the chances of getting something like the following to work?
```
var block0 = document.querySelector('#block0');
block0.setAttribute('position', '2 1 3');
```

Have a great weekend
Jasper

**Comment by diarmidmackenzie** [comment]
Hi Jasper, thanks for the question.

First, I wanted to clarify how you'd make your example work with the current implementation.  It just requires one additional line of code:

```
var block0 = document.querySelector('#block0');
block0.setAttribute('position', '2 1 3');
block0.emit('object3DUpdated );
```

You could also do this (following A-Frame guidance for updating position [here](https://aframe.io/docs/1.3.0/components/position.html#updating-position):
```
var block0 = document.querySelector('#block0');
block0.object3D.position.set(2, 1, 3);
block0.emit('object3DUpdated );
```

Fundamentally the THREE.js InstancedMesh object renders positions based on an array of matrices, rather than the object3D positions themselves, so there needs to be some trigger to the instanced-mesh component that prompts it to update the matrices in this array.

There are a few alternatives to requiring the object3DUpdated event, that I have considered, but none of them is ideal.

1. Trigger on [componentchanged](https://aframe.io/docs/1.3.0/core/entity.html#events) event

There are several issues with this:
- It won't catch the case where the object3D is updated directly (per my 2nd example above), rather than by modifying the position attribute.  Given that this is how A-Frame *recommends* object positions are updated, that's quite an issue.
- It won't catch the case where an object moves relative to te instanced mesh because a parent objects is moved, scaled or rotated.
- It fires for *any* change to any component on the entity, which might be a performance concern.


2. Run a tick() to check for object movements.

This is unlikely to be a good solution, for performance reasons:
- If the tick doesn't run every frame (e.g. only every 100msecs), then you'll get lag in object positions.  That might be acceptable in some contexts, but most of the time that's going to look pretty poor visually.
- If the tick runs every frame, it's probably more performant just to re-copy all the object matrices every frame (see 3.) since that involves less computation than checking for changes to matrices.



3. Re-copy every object transform into the instanced mesh matrix array every frame.

This is probably the most viable alternative.

I didn't take this approach for performance reasons, particularly given that instanced-mesh specifically targets large numbers of instanced mesh members.  However there may be use cases where the number of instanced mesh members is small enough that this approach is reasonable (but still large enough that it's worthwhile having an instanced mesh).

With 10s of members this approach would probably be fine.

Simplest solution would be to just update all matrices every frame on tick().

I think this should just come down to doing the work in `memberModified()` for every object on every frame.  Small bit of refactoring to `memberModified()` needed, as it currently assumes it has an event supplied as an argument.

I'd be willing to add this as an option to the instanced-mesh component.  I can probably get to this in the next couple of weeks. 
 I'd also happy to take a PR if you wanted to have a go at this.

Aside from matrix changes, object3DUpdated is also used to signal changes in visibility.  Some thought needed in terms of handling visibility changes in a similar way - but I could spin off a separate issue to track that.

**Comment by aframo** [comment]
Thank you for the detailed explanation. 

I was able to use 'object3DUpdated' and remix the default A-Frame position component into something rudimentary that changes the position of instanced entities.

`AFRAME.registerComponent('im-position', {
  schema: {type: 'vec3'},

  update: function () {
    var object3D = this.el.object3D;
    var data = this.data;
    object3D.position.set(data.x, data.y, data.z);

    // Emit updated object3D position for instanced-mesh
    this.el.emit('object3DUpdated');
  }
(...)
`
(the "im-" in the name refers to "instanced mesh" and I love how it spells to sound like it's _imposing_)

The component only works with slow and consecutive changes, but helped me understand the performance bottlenecks you described a bit. Essentially, the THREE.js InstancedMesh object becomes too busy adjusting the matrices for the updates it receives to keep up with changes when there's a lot of it? Or at least that's the best I can make sense of the lag I'm witnessing, even though it doesn't explain why initializing the meshes is instantaneous... 

I'd love to have a go at this, albeit I'm a bit out of my element with InstancedMesh.
Wouldn't the best approach here be to have a component (say, `im-position`) cumulate the updates between frames and then release the batch (array?) to the InstancedMesh object?

[ 2 hours later ]

After poking around a bit in instanced-mesh, I think I'm understanding that `mesh.setMatrixAt(index, this.componentMatrix)` adds the positions of the meshes into the InstancedMesh object one by one, index by index. It would be great if there was a way to do this in bulk. Any chance it's possible to put instanced meshes inside other instanced meshes and play around that angle to have a single index cause changes in many of the meshes?

I'm just throwing around wild ideas now because it's all a bit much to wrap my head around in one day. I'll play around with InstancedMesh some more tomorrow and let you know if I find anything useful.

**Comment by diarmidmackenzie** [comment]
Hi - it would be good to understand your use case a bit better.

It sounds like you have a large instanced mesh, and within that you have "collections" of members, where you'd like to be able to efficiently apply a transform to all members in the collection?

Currently implementation doesn't handle this well.  Currently available approaches would be:
- if collections are small, just fire object3DUpdated for each member
- if collections are large, use a separate instanced-mesh per collection.

In theory, I think nested instanced meshes should be possible (i.e. you could make the code work), but it's not clear to me whether there are any situations where that approach would make a meaningful positive difference to performance vs. the already-available options.

It would also be helpful to see the tests you have written, where you see performance slowdown.  How does performance compare to this example?
https://diarmidmackenzie.github.io/instanced-mesh/tests/rotation-high-scale.html

**Comment by aframo** [comment]
Yeah, the rotation example doesn't work with my use case (changing positions) because that just gives all the members the same attribute (in my case position, clumping them up). Doing the updates outside of a for-loop has dreadful performance.

[https://aframo.github.io/instanced-mesh/tests/im-position.html](https://aframo.github.io/instanced-mesh/tests/im-position.html)
Essentially, I'd love to have something like this with... 100x better performance? I'm starting to feel like I'm asking for something labyrinthine (might even need big changes in Three.js?). How do triple-A game engines animate hundreds of objects so gracefully?

**Comment by jaspermolgvits** [comment]
I just realized I can simply swap between animation components (which are ways faster) to have a variety of animations without turning it into a Monte Carlo fluid simulation haha...

Sorry for my confusion, and thank you for your tips. Next time I'll start off with a live example of what I'm trying to do.

I'm not sure changing the position attribute on a DOM element could ever be performant, so this issue is probably closed for now?

**Comment by aframo** [comment]
I just realized I can simply swap between animation components (which are ways faster) to have a variety of animations without turning it into a Monte Carlo fluid simulation haha...

Sorry for my confusion, and thank you for your tips. Next time I'll start off with a live example of what I'm trying to do.

I'm not sure changing the position attribute on a DOM element could ever be performant, so this issue is probably closed for now?

**Comment by diarmidmackenzie** [comment]
Hi, sorry I didn't get a chance to look at your code yet.  I will try to do so.

I still don't understand your use case, but I'm wondering if it might be better met by using a particle system rather than an instanced mesh?

See e.g.
https://github.com/IdeaSpaceVR/aframe-particle-system-component

**Comment by aframo** [comment]
Aren't particle systems only for 2D sprites?

I'm hoping to create a little story game over the summer and I'm just testing ways to make it look (and perform) decent. Using particle systems is a good idea, and I'll definitely try it eventually, but right now I'm just trying to figure out the 3D side of things.

No worries about looking at my code.

**Comment by diarmidmackenzie** [comment]
Took a look at your code.  A few comments.

BULK = false, so you are only moving one block each 10msec cycle.  Is that what you intended.

Comment says that when BULK is true, all blocks end up in the same place.  That's because you aren't changing m or p inside your i and j loops.

You will get better performance by using el.object3D.position.set() rather than setAttribute.

Finally, writing code that runs on documentReady goes against A-Frame guidelines.  A-Frame recommends that all JS code is wrapped inside a component and applied to a suitable entity (which can be the a-scene).  Rather than using setTimeout, better to do your updates on a tick() and move a variable number of blocks depending on the time delta since the last tick().

**Comment by kylebakerio** [comment]
> There are a few alternatives to requiring the object3DUpdated event, that I have considered, but none of them is ideal.

Just wanted to propose another possible candidate, though I think the curren situation is perfectly acceptable:

4. make a function available that wraps the update +  emit the event.

I don't think you should take the existing option away, of course, as if I want to do updates to 20 items and only then emit the event, that should be available to me--but for basic use cases of moving individual elements, this might be ideal.

Perhaps the best place for that function to live is in the `update()` call of the individual elements? have `el.setAttribute('instance-mesh', 'position', '1 1 1')` available as an option, along with rotation and scale; this would get caught by `update()`, which would set the object3d and then emit the event. Like A-Frame, you can recommend that they do the manipulation directly if performance sensitive, but otherwise this convenience wrapper is there if they want it.

Could also make it so that so those same changes could be made to all members at once if changes are made on a property of the component of the "primary instance".

Just random thoughts I had reading through on this while using it for the first time, great work, went super smoothly!

**Comment by diarmidmackenzie** [comment]
Thanks, Kyle.

Have been thinking about how physics and instancing work together, and my view is it would be useful to have a config option on an instanced entity where it automatically refreshes position every frame, rather than waiting for the object3Dupdated event.

I think that will give better performance in situations where you expect objects to move a lot.

Firing an event for every single object every frame is a big overhead.

Planning to do this along with some instancing+physics examples in the fairly near future.

**Comment by kylebakerio** [comment]
instancing + physics will be a potent combo, I look forward to that.

**Comment by diarmidmackenzie** [comment]
Fixed by #13
